### PR TITLE
Support for disabling PLT for better function call performance

### PR DESCRIPTION
**Pull request by GabrielMajeri**
This PR gives `rustc` the ability to skip the PLT when generating function calls into shared libraries. This can improve performance by reducing branch indirection.

AFAIK, the only advantage of using the PLT is to allow for ELF lazy binding. However, since Rust already [enables full relro for security](https://github.com/rust-lang/rust/pull/43170), lazy binding was disabled anyway.

This is a little known feature which is supported by [GCC](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html) and [Clang](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fplt) as `-fno-plt` (some Linux distros [enable it by default](https://git.archlinux.org/svntogit/packages.git/tree/trunk/makepkg.conf?h=packages/pacman#n40) for all builds).

Implementation inspired by [this patch](https://reviews.llvm.org/D39079#change-YvkpNDlMs_LT) which adds `-fno-plt` support to Clang.

## Performance

I didn't run a lot of benchmarks, but these are the results on my machine for a `clap` [benchmark](https://github.com/clap-rs/clap/blob/master/benches/05_ripgrep.rs):

```
 name              control ns/iter  no-plt ns/iter  diff ns/iter  diff %  speedup 
 build_app_long    11,097           10,733                  -364  -3.28%   x 1.03 
 build_app_short   11,089           10,742                  -347  -3.13%   x 1.03 
 build_help_long   186,835          182,713               -4,122  -2.21%   x 1.02 
 build_help_short  80,949           78,455                -2,494  -3.08%   x 1.03 
 parse_clean       12,385           12,044                  -341  -2.75%   x 1.03 
 parse_complex     19,438           19,017                  -421  -2.17%   x 1.02 
 parse_lots        431,493          421,421              -10,072  -2.33%   x 1.02 
```

A small performance improvement across the board, with no downsides. It's likely binaries which make a lot of function calls into dynamic libraries could see even more improvements. [This comment](https://patchwork.ozlabs.org/patch/468993/#1028255) suggests that, in some cases, `-fno-plt` could improve PIC/PIE code performance by 10%.

## Security benefits

**Bonus**: some of the speculative execution attacks rely on the PLT, by disabling it we reduce a big attack surface and reduce the need for [`retpoline`](https://reviews.llvm.org/D41723).

## Remaining PLT calls

The compiled binaries still have plenty of PLT calls, coming from C/C++ libraries. Building dependencies with `CFLAGS=-fno-plt CXXFLAGS=-fno-plt` removes them.

### CODE REVIEW

**Comment by eddyb** [comment]
cc @alexcrichton

**Comment by cuviper** [comment]
> AFAIK, the only advantage of using the PLT is to allow for ELF lazy binding. However, since Rust already [enables full relro for security](https://github.com/rust-lang/rust/pull/43170), lazy binding was disabled anyway.

Note that PPC64 is only defaulting to partial relro due to an old ld.so bug in bind-now.
https://github.com/rust-lang/rust/pull/43170/files#diff-b2d51315427bd679ca33d47167e82171R20

There's also an option for `-Z relro-level={full,partial,off}`.  I'll try to see if I see similar PPC64 issues arise with `-fno-plt`, but my initial feeling is that we should only enable this in conjunction with `relro-level=full`.

**Comment by nagisa** [comment]
Please add a flag that controls this behaviour. For now it can be a debug `-Z` flag, similar to other such flags (e.g. `-Zmutable-noalias`).

> Also, I'm not sure how to fix the failing test. Is it possible for CHECK: Function Attrs to allow extra attributes in the output, besides the ones being tested?

CHECK lines only check for matching line prefix. That test in particular seems to be testing for `naked` only, in which case you can probably remove the other attribute from the `CHECK` line to have it pass. Alternatively, you might have some success with [pattern matching syntax](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-pattern-matching-syntax).

**Comment by nagisa** [comment]
> objdump says that the number of function calls using the PLT goes down, but a lot of functions calls are still using the PLT. So I'm guessing I have to add this in other places too, but I'm not very familiar with the code.

A large number of `@PLT` symbols likely come from outside the rust ecosystem (e.g. glibc, llvm, etc.). Those might need to be taken care of independently (by changing build system configuration, perhaps?). You might want to submit a similar patch to the [cc crate](https://github.com/alexcrichton/cc-rs).

---

(Addressed to not-author, but somebody who knows how to do perf runs) I also think a perf run would be great, but not sure how to start it.

**Comment by varkor** [comment]
@rust-timer build https://github.com/rust-lang/rust/commit/57476313f10d8c8068ce4241d959e228fc7eb6a3

**Comment by GabrielMajeri** [comment]
Perf results are in, nice improvements on wall time. From what I've seen, the patch currently only removes *about 20%* of the total PLT calls, there's probably still some more performance to be gained. 

@nagisa 
> Please add a flag that controls this behaviour. 

Alright, I've implemented a `-Z no-plt` flag, but should this option be enabled or disabled by default?

Also, what should we do if the user specified `-Z no-plt`, but the flag is then ignored, because full relro isn't supported (for example, as cuviper mentioned, linker issues on PowerPC64)?

> symbols likely come from outside the rust ecosystem

Thanks for the tip, but I'm still unable to get rid of the PLT. I've added `CFLAGS=-fno-plt` to my system, and rebuilt the compiler from source, but `rustc` still generates _lots_ of calls which use the PLT.

If I build C binaries on my system, the final binary doesn't even have a `.plt` section, it is completly removed.

**Comment by cuviper** [comment]
> Alright, I've implemented a `-Z no-plt` flag, but should this option be enabled or disabled by default?

I think the way you've documented it is fine, "(default: PLT is disabled if full relro is enabled)".

> Also, what should we do if the user specified `-Z no-plt`, but the flag is then ignored, because full relro isn't supported (for example, as cuviper mentioned, linker issues on PowerPC64)?

Don't ignore it.  These are advanced options -- if the user asks for `plt=off` without full relro, let them deal with the implications.  So the check is something like `plt.unwrap_or(relro != Full)`.

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -1387,6 +1387,10 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
           "output a json file with profiler results"),
     emit_stack_sizes: bool = (false, parse_bool, [UNTRACKED],
           "emits a section containing stack size metadata"),
+    plt: Option<bool> = (Some(false), parse_opt_bool, [TRACKED],
```
Comment (cuviper) [inline src/librustc/session/config.rs]: Shouldn't this be initialized `None`, so we know whether it has been explicitly set?

**Comment by joshtriplett** [comment]
This looks great to me; it's a good optimization with no downsides.

**Comment by cuviper** [comment]
BTW, one case where you do need PLT is to use tools like [ltrace](http://man7.org/linux/man-pages/man1/ltrace.1.html) and [latrace](http://people.redhat.com/jolsa/latrace/latrace.html) which hook PLT calls.  But I think they're uncommon enough that it's OK for us to take the performance win by default, and have the `-Z plt` option when you really want to use those tools.

File: src/librustc/session/config.rs (hunk 1 above)
Comment (nagisa) [reply src/librustc/session/config.rs]: Yes, this should be a `None`. While `Some(false)` makes a long of sense with the current defaults, it is easier to make changes to defaults and code overall, when there are no surprises between “not provided” and “`-Zplt=no` is provided”.

**Comment by nagisa** [comment]
Is it possible to warn if the options are conflicting e.g. `-Zplt=yes` and relro is not full? It would be confusing to have `-Zplt=yes` silently ignored.

**Comment by nagisa** [comment]
Or rather, I should be asking what is the failure mode when `-Zplt=yes` and “relro” is not “full”? I see that the code only takes relro into account only when PLT usage is not forced one way or the other.

**Comment by nagisa** [comment]
Also, sorry to be making so many comments at once, but if you ping me on IRC/Zulip (same nick as here on both), we could work together on figuring out where the rest of the PLT calls come from.

**Comment by nagisa** [comment]
Doing another perf run – there is now way fewer `@PLT` calls since the last run.

@bors try

**Comment by alexcrichton** [comment]
Hm this'll probably need a force push to re-try unfortunately

**Comment by GabrielMajeri** [comment]
@alexcrichton I've rebased and force-updated.

**Comment by nagisa** [comment]
@rust-timer build f4a266585ffc702dfdbfafdec11e09f6128d6c38

**Comment by Valinora** [comment]
Perf is in, results seem mixed but mostly positive

**Comment by nagisa** [comment]
r=me on the implementation, just waiting for the checkboxes.

**Comment by nagisa** [comment]
@GabrielMajeri if you’re around it would be great if you could squash commits a little. “Fix failing test” or “rename flag” have very little value as separate commits.

**Comment by GabrielMajeri** [comment]
Well I don't know what's wrong with `gnux32`, but in my defense [Clang  also fails to compile](https://gcc.godbolt.org/z/ALEsEJ) when I pass `-fPIC -fno-plt` on the `gnux32` target.
And it's [apparently a know LLVM bug](https://bugs.llvm.org/show_bug.cgi?id=36743).

Should we just blacklist gnux32 for the  `-Z plt` option?

**Comment by nagisa** [comment]
Simply ignoring the test for the target should be fine.

**Comment by cuviper** [comment]
> Simply ignoring the test for the target should be fine.

Wouldn't that still mean the _default_ no-plt is broken on gnux32?

**Comment by nagisa** [comment]
Oh right, I neglected to remember we’re making `-Zplt=off` by default.

In that case the default should be indicated in the target specification and the `gnux32` should explicitly enable plt by default.

Thinking about it now, indicating PLT default in the target specification is a good idea nevertheless, but unlike with the flag target specifications aren’t covered by the stability, and would become insta-stable…

**Comment by GabrielMajeri** [comment]
The way I see it, `-Z plt=off` is an optimization, we don't guarantee it does anything (it's a best effort kind of thing). For now, I changed the code to unconditionally disable the optimization on `gnux32`, at least until LLVM gets fixed.

**Comment by nagisa** [comment]
This should still be a part of target specification because there are
targwta defined outside of the compiler and we do t test them all.

**Comment by GabrielMajeri** [comment]
@nagisa As far as I understand, even if somebody defines some external target with this ABI, there shouldn't be an issue.

The code checks for the (custom) target's `llvm-target` attribute to see if it contains `gnux32`. This is as far as we know the only ABI where LLVM currently has an issue (due to a bug).

For example, Clang accepts this option for all targets and ABIs (even Windows). On targets where it doesn't do anything, it emits the attributes, and LLVM just ignores them (except for this buggy ABI which crashes).

**Comment by nagisa** [comment]
I'm arguing that the author of external target specifications should have a
full control over such matters.

While gnux32 is the only target we know that crashes, our triplet coverage
is incomplete and we can't say for sure that any future triplets llvm may
support will be problem free either.

In the past platform specific settings and workarounds were always stored
within the target files and I see no reason whatsoever for us to change
that now. And it is strictly more flexible than checking whether llvm
triple string ends in a specific substring.

I also don't see much of an issue with allowing people to force arbitrary
configurations (i.e. allowing them to specify -Zplt=off) even if we know
they might not be working, ignored, irrelevant etc. At worst they'll
encounter an ICE. In fact, if they do explicitly request for some
behaviour, I believe that should override any workaround or default (which
is not what happens with the current iteration of the PR)

**Comment by pnkfelix** [comment]
(@nagisa said at T-compiler meeting that we can un-nominate this)

**Comment by GabrielMajeri** [comment]
@nagisa ok, I've added a `needs_plt` target option which can be customized for each target. It is used to help determine a default for the PLT option (and `-Z plt` always overrides the setting).

**Comment by alexcrichton** [comment]
FWIW this looks like it may cause bugs in LLVM on i686-unknown-linux-gnu. I noticed that [stdsimd's CI](https://travis-ci.org/rust-lang-nursery/stdsimd) was [failing](https://travis-ci.org/rust-lang-nursery/stdsimd/jobs/443791815) for i686-unknown-linux-gnu because rustc was segfaulting. Some local investigation showed a segfault in LLVM. We compile that target with `-C relocation-model=static` on CI, and I believe the combination of 32-bit Linux with `-C relocation-model=static` was causing the issue. I haven't had a chance to dig deeper. I've worked around it with `-Z plt=yes`

**Comment by eddyb** [comment]
FWIW, this broke pretty badly on certain (older) distro toolchains, but because `dylib` is largely unused, it took someone (ab)using `proc_macro::bridge` to run proc macros outside of `rustc` to trigger it: #61539.

**Comment by MaskRay** [comment]
I think `-Z plt=no` is not a good default. Sent https://github.com/rust-lang/rust/pull/106380 to disable it. (Thanks to @GabrielMajeri for mentioning this review as I cannot find it with the commits).

If you see positive benchmark results, it is likely because dynamically linked libc calls dominate. If one statically links libc, `-Z plt=no` is going to be a pessimization.
Plus, x86-32 requires very new lld (as a maintainer, I just added support for `___tls_get_addr`; older lld will create a silently corrupted executable).
For GNU ld, a relatively new one is needed: 2016-06 https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=e2cbcd9156d1606a9f2153aecd93a89fe6e29180 (and a counterpart for x86-32)

**Pull request merged**